File: 486/solution.py
class Solution:
    def predictTheWinner(self, nums: List[int]) -> bool:

        def dfs(left, right):
            if left==right:
                return nums[left]
            leftSide = nums[left] - dfs(left+1, right)
            rightSide = nums[right] - dfs(left, right-1)

            return max(leftSide, rightSide)
        res = dfs(0, len(nums)-1)>=0
        return res

File: 486/solution.java
class Solution {

    public int dfs(int[] arr, int left, int right){
        if(left==right)
            return arr[left];
        int leftSide = arr[left] - dfs(arr, left+1, right);
        int rightSide = arr[right] - dfs(arr, left, right-1);
        return Math.max(leftSide, rightSide);
    }
    public boolean predictTheWinner(int[] nums) {
        return dfs(nums, 0, nums.length-1)>=0;
    }
}
